demo/runtime/src/dispatch.rs

```diff
@@ -16,79 +16,228 @@
 
 //! Dispatch system. Just dispatches calls.
 
-use demo_primitives::{Function, Proposal, AccountId};
-use runtime::{staking, system, session, democracy, council, council_vote, timestamp};
-
-/// Dispatch a proposal.
-pub fn proposal(proposal: Proposal) {
-	match proposal {
-		Proposal::SystemSetCode(ref a) =>
-			system::privileged::set_code(a),
-		Proposal::SessionSetLength(a) =>
-			session::privileged::set_length(a),
-		Proposal::SessionForceNewSession =>
-			session::privileged::force_new_session(),
-		Proposal::StakingSetSessionsPerEra(a) =>
-			staking::privileged::set_sessions_per_era(a),
-		Proposal::StakingSetBondingDuration(a) =>
-			staking::privileged::set_bonding_duration(a),
-		Proposal::StakingSetValidatorCount(a) =>
-			staking::privileged::set_validator_count(a),
-		Proposal::StakingForceNewEra =>
-			staking::privileged::force_new_era(),
-		Proposal::DemocracyCancelReferendum(a) =>
-			democracy::privileged::cancel_referendum(a),
-		Proposal::DemocracyStartReferendum(a, b) =>
-			democracy::privileged::start_referendum(*a, b),
-		Proposal::CouncilSetDesiredSeats(a) =>
-			council::privileged::set_desired_seats(a),
-		Proposal::CouncilRemoveMember(a) =>
-			council::privileged::remove_member(&a),
-		Proposal::CouncilSetPresentationDuration(a) =>
-			council::privileged::set_presentation_duration(a),
-		Proposal::CouncilSetTermDuration(a) =>
-			council::privileged::set_term_duration(a),
-		Proposal::CouncilVoteSetCooloffPeriod(a) =>
-			council_vote::privileged::set_cooloff_period(a),
-		Proposal::CouncilVoteSetVotingPeriod(a) =>
-			council_vote::privileged::set_voting_period(a),
+use runtime::{staking, system};
+pub use rstd::prelude::Vec;
+pub use codec::{Slicable, Input, NonTrivialSlicable};
+
+#[macro_export]
+macro_rules! impl_dispatch {
+	(
+		pub mod $mod_name:ident;
+		$(
+			fn $fn_name:ident(
+				$(
+					$param_name:ident : $param:ty
+				),*
+			)
+			= $id:expr ;
+		)*
+	) => {
+		pub mod $mod_name {
+			use super::*;
+
+			#[derive(Clone, Copy, PartialEq, Eq)]
+			#[cfg_attr(feature = "std", derive(Serialize, Deserialize, Debug))]
+			#[repr(u32)]
+			#[allow(non_camel_case_types)]
+			enum Id {
+				$(
+					#[allow(non_camel_case_types)]
+					$fn_name = $id,
+				)*
+			}
+
+			impl Id {
+				/// Derive `Some` value from a `u8`, or `None` if it's invalid.
+				fn from_u8(value: u8) -> Option<Id> {
+					use self::*;
+					let functions = [
+						$( Id::$fn_name ),*
```

Comment:
why not generate a match statement?

Reply:
fixed (not intentional - just was adapting existing code)